File: app/src/main/java/sabalan/paydar/mohtasham/ezibazi/api_service/main/ShopService.java
package sabalan.paydar.mohtasham.ezibazi.api_service.main;

import android.content.Context;


import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import sabalan.paydar.mohtasham.ezibazi.api_service.ApiRequest;
import sabalan.paydar.mohtasham.ezibazi.api_service.Urls;
import sabalan.paydar.mohtasham.ezibazi.system.pref_manager.CityPrefManager;
import sabalan.paydar.mohtasham.ezibazi.model.Game;
import sabalan.paydar.mohtasham.ezibazi.model.Paginate;

public class ShopService {
  private Context context;
  public ShopService(Context context){
    this.context = context;
  }


  public void getMainShops(final int page_num, final onShopsReceived onShopsReceived){
    String url = Urls.SHOP_INDEX0 + "/"+ Integer.toString(new CityPrefManager(context).getCityId()) + "?page=" + page_num;
    ApiRequest apiRequest = new ApiRequest(context);
    apiRequest.request(ApiRequest.GET, url, null, false, new ApiRequest.onDataReceived() {
      @Override
      public void onReceived(JSONObject response, int status, String message, boolean error) {
        if (error){
          onShopsReceived.onReceived(status, message, new ArrayList<Game>(), new Paginate());
          return;
        }

        try {
          List<Game> games =  new ArrayList<>();
          JSONObject jsonData = response.getJSONObject("data");
          Paginate paginate = Paginate.Parser.parse(jsonData);
          JSONArray data = jsonData.getJSONArray("data");
          for (int i=0 ; i<data.length() ; i++){
            JSONObject gameObj = data.getJSONObject(i);
            games.add( Game.Parser.parse(gameObj));
          }

          onShopsReceived.onReceived(status, message, games, paginate);
        } catch (JSONException e) {
          e.printStackTrace();
        }

      }
    });


  }


  public void getSearchedShops(JSONObject obj, final onSearchedShopsReceived onSearchedShopsReceived){
    ApiRequest apiRequest = new ApiRequest(context);
    apiRequest.request(ApiRequest.POST, Urls.SHOP_SEARCH, obj, false, new ApiRequest.onDataReceived() {
      @Override
      public void onReceived(JSONObject response, int status, String message, boolean error) {
        if (error){
          onSearchedShopsReceived.onReceived(status, message, new ArrayList<Game>());
          return;
        }

        try {
          List<Game> games =  new ArrayList<>();
          if(status == 1) {
            JSONArray data = response.getJSONArray("data");
            for (int i = 0; i < data.length(); i++) {
              games.add(Game.Parser.parse(data.getJSONObject(i)));
            }
          }

          onSearchedShopsReceived.onReceived(status, message, games);
        } catch (JSONException e) {
          e.printStackTrace();
        }

      }
    });


  }


  public void getRelatedShops(final int game_id, final onRelatedShopsReceived onRelatedShopsReceived){
    String url = Urls.GAME_RELATED_GAME_FOR_SHOP + "/"+Integer.toString(game_id) ;
    ApiRequest apiRequest = new ApiRequest(context);
    apiRequest.request(ApiRequest.GET, url, null, false, new ApiRequest.onDataReceived() {
      @Override
      public void onReceived(JSONObject response, int status, String message, boolean error) {
        if (error){
          onRelatedShopsReceived.onReceived(status, message, new ArrayList<Game>());
          return;
        }

        try {
          List<Game> games =  new ArrayList<>();
          JSONArray data = response.getJSONArray("data");
          for(int i=0 ; i<data.length();i++){
            games.add(Game.Parser.parse(data.getJSONObject(i)));
          }

          onRelatedShopsReceived.onReceived(status, message, games);
        } catch (JSONException e) {
          e.printStackTrace();
        }
      }
    });

  }


  public void getSpecialShop(int id, final onSpecialShopReceived onSpecialshopReceived){
    String url = Urls.SHOP_INDEX + "/" + Integer.toString(id);
    ApiRequest apiRequest = new ApiRequest(context);
    apiRequest.request(ApiRequest.GET, url, null, false, new ApiRequest.onDataReceived() {
      @Override
      public void onReceived(JSONObject response, int status, String message, boolean error) {
        if (error){
          onSpecialshopReceived.onReceived(status, message, new Game());
          return;
        }

        Game game = new Game();
        try {
          JSONObject gameObj = response.getJSONObject("data");
          game = Game.Parser.parse(gameObj);

          onSpecialshopReceived.onReceived(status, message, game);
        } catch (JSONException e) {
          e.printStackTrace();
        }
      }
    });


  }


  public interface onShopsReceived{
    void onReceived(int status, String message, List<Game> games, Paginate paginate);
  }

  public interface onSearchedShopsReceived{
    void onReceived(int status, String message, List<Game> games);
  }

  public interface onRelatedShopsReceived{
    void onReceived(int status, String message, List<Game> games);
  }

  public interface onSpecialShopReceived{
    void onReceived(int status, String message, Game game);
  }


}

File: app/src/main/java/sabalan/paydar/mohtasham/ezibazi/api_service/main/ShopService.kt
package sabalan.paydar.mohtasham.ezibazi.api_service.main

import android.content.Context


import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

import sabalan.paydar.mohtasham.ezibazi.api_service.ApiRequest
import sabalan.paydar.mohtasham.ezibazi.api_service.Urls
import sabalan.paydar.mohtasham.ezibazi.system.pref_manager.CityPrefManager
import sabalan.paydar.mohtasham.ezibazi.model.Game
import sabalan.paydar.mohtasham.ezibazi.model.Paginate

class ShopService(private val context: Context) {


    fun getMainShops(page_num: Int, onShopsReceived: onShopsReceived) {
        val url = Urls.SHOP_INDEX0 + "/" + Integer.toString(CityPrefManager(context).cityId) + "?page=" + page_num
        val apiRequest = ApiRequest(context)
        apiRequest.request(ApiRequest.GET, url, null!!, false, object : ApiRequest.onDataReceived {
            override fun onReceived(response: JSONObject?, status: Int, message: String, error: Boolean) {
                if (error) {
                    onShopsReceived.onReceived(status, message, ArrayList<Game>(), Paginate())
                    return
                }

                try {
                    val games = ArrayList<Game>()
                    val jsonData = response!!.getJSONObject("data")
                    val paginate = Paginate.Parser.parse(jsonData)
                    val data = jsonData.getJSONArray("data")
                    for (i in 0 until data.length()) {
                        val gameObj = data.getJSONObject(i)
                        games.add(Game.Parser.parse(gameObj))
                    }

                    onShopsReceived.onReceived(status, message, games, paginate)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

            }
        })


    }


    fun getSearchedShops(obj: JSONObject, onSearchedShopsReceived: onSearchedShopsReceived) {
        val apiRequest = ApiRequest(context)
        apiRequest.request(ApiRequest.POST, Urls.SHOP_SEARCH, obj, false, object : ApiRequest.onDataReceived {
            override fun onReceived(response: JSONObject?, status: Int, message: String, error: Boolean) {
                if (error) {
                    onSearchedShopsReceived.onReceived(status, message, ArrayList())
                    return
                }

                try {
                    val games = ArrayList<Game>()
                    if (status == 1) {
                        val data = response!!.getJSONArray("data")
                        for (i in 0 until data.length()) {
                            games.add(Game.Parser.parse(data.getJSONObject(i)))
                        }
                    }

                    onSearchedShopsReceived.onReceived(status, message, games)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

            }
        })


    }


    fun getRelatedShops(game_id: Int, onRelatedShopsReceived: onRelatedShopsReceived) {
        val url = Urls.GAME_RELATED_GAME_FOR_SHOP + "/" + Integer.toString(game_id)
        val apiRequest = ApiRequest(context)
        apiRequest.request(ApiRequest.GET, url, null!!, false, object : ApiRequest.onDataReceived {
            override fun onReceived(response: JSONObject?, status: Int, message: String, error: Boolean) {
                if (error) {
                    onRelatedShopsReceived.onReceived(status, message, ArrayList<Game>())
                    return
                }

                try {
                    val games = ArrayList<Game>()
                    val data = response!!.getJSONArray("data")
                    for (i in 0 until data.length()) {
                        games.add(Game.Parser.parse(data.getJSONObject(i)))
                    }

                    onRelatedShopsReceived.onReceived(status, message, games)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

            }
        })

    }


    fun getSpecialShop(id: Int, onSpecialshopReceived: onSpecialShopReceived) {
        val url = Urls.SHOP_INDEX + "/" + Integer.toString(id)
        val apiRequest = ApiRequest(context)
        apiRequest.request(ApiRequest.GET, url, null!!, false, object : ApiRequest.onDataReceived {
            override fun onReceived(response: JSONObject?, status: Int, message: String, error: Boolean) {
                if (error) {
                    onSpecialshopReceived.onReceived(status, message, Game())
                    return
                }

                var game = Game()
                try {
                    val gameObj = response!!.getJSONObject("data")
                    game = Game.Parser.parse(gameObj)

                    onSpecialshopReceived.onReceived(status, message, game)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

            }
        })


    }


    interface onShopsReceived {
        fun onReceived(status: Int, message: String, games: List<Game>, paginate: Paginate)
    }

    interface onSearchedShopsReceived {
        fun onReceived(status: Int, message: String, games: List<Game>)
    }

    interface onRelatedShopsReceived {
        fun onReceived(status: Int, message: String, games: List<Game>)
    }

    interface onSpecialShopReceived {
        fun onReceived(status: Int, message: String, game: Game)
    }


}
